api/v1/services/ProjectService.js

```diff
@@ -0,0 +1,208 @@
+ * @param  {Object} Contains name, description, repo, and is_published
+ * @return {Promise} resolving to the newly-created project
+ * @throws InvalidParameterError when a project exists with the specified name
+ */
+module.exports.createProject = function (attributes) {
+	if(typeof attributes.repo === 'undefined'){
+		attributes.description = '';
+	}
+	if(typeof attributes.isPublished === 'undefined'){
+		attributes.isPublished = 0; //false
+	}
+
+	validationName = attributes.name.toLowerCase();
+
+	var project = Project.forge(attributes);
+	return project
+		.validate()
+		.catch(Checkit.Error, utils.errors.handleValidationError)
+		.then(function (validated) {
+			return Project.findByName(validationName);
+		})
+		.then(function (result){
+			if (!_.isNull(result)) {
+				var message = "A project with the given name already exists";
+				var source = "name";
+				throw new errors.InvalidParameterError(message, source);
+			}
+
+			return project
+				.save()
+				.then(function (project) {
+					return project;
+				});
+		})
+}
+
+/**
+ * Returns a project with the specified project id
+ * @param  {int} ID of the project
+ * @return {Promise} resolving to the project
+ * @throws InvalidParameterError when a project doesn't exist with the specified ID
+ */
+module.exports.findProjectById = function (id) {
+	return Project
+		.findById(id)
+		.then(function (result) {
+			if(_.isNull(result)){
+				var message = "A project with the given ID cannot be found";
+				var source = "id";
+				throw new errors.NotFoundError(message, source);
+			}
+
+			return _Promise.resolve(result);
+		});
+}
+
+/**
+ * Update a key value pair in a project
+ * @param  {Project} Project that will be updated
+ * @param  {String} Key is the name of the attribute
+ * @param  {String} Value is the new value of the attribute
+ * @return {Promise} resolving to the updated project
+ * @throws InvalidParameterError when the key is not valid
+ */
+module.exports.updateProject = function (project, key, value) {
+	if(projectKeys.indexOf(key.toLowerCase()) == -1){
+		var message = "The given key is invalid";
+		var source = "key";
+		throw new errors.InvalidParameterError(message, source);
+	}
+	return project
+		.set(key, value)
+		.validate()
+		.catch(Checkit.Error, utils.errors.handleValidationError)
+		.then(function (validated) {
+			return project.save();
+		});
+}
+
+/**
+ * Helper function for determining valid project/mentor ids
+ * @param  {Int} ID of the project assigned to the mentor
+ * @param  {Int} ID of the mentor assigned to the project
+ * @return {Promise} resolving to whether or not the ids are valid
+ * @throws InvalidParameterError when a project or mentor doesn't exist with the specified ID
+ */
+_isProjectMentorValid = function (project_id, mentor_id) {
+	return Project
+		.findById(project_id)
+		.then(function (result) {
+			if(!_.isNull(result)) {
+				return _Promise.resolve(true);
+			}else{
+				return _Promise.resolve(false);
```

Comment:
How about you just throw an error here and then have a catch block that handles it somewhere else? That way you won't be resolving a value into a portion of the pipeline when you don't actually want to use it.